CONTRIBUTING.md

```diff
@@ -1,7 +1,75 @@
 # Contribution Guidelines
 
 We welcome issues and pull requests at https://github.com/NixOS/nixfmt.
+PRs that change the format should preferably be discussed in an issue first.
 
-We use stylish-haskell to format our code. Make sure to format your
-contributions with it. It is available within the `nix-shell` and it should pick
-up the configuration from the repository automatically.
+When contributing, please try to familiarize yourself with the [Nix Format Standard](https://github.com/NixOS/nixfmt/blob/master/standard.md), because the distinction between bugs in Nixfmt and issues with the format is relevant in many cases.
+
+You can also reach us on Matrix at `#nix-formatting:nixos.org`.
+
+## Contributing code
+
+Most changes to the format are going to be implemented in `Pretty.hs`.
+Sometimes, other areas of the code base need fixing (CLI, Parser, etc.) as well.
+However, if you find yourself touching `Predoc.hs`, please open an issue first or contact us on Matrix.
+That file contains the IR and the renderer, they are brittle and require expertise.
+
+### Testing
+
+There are no unit tests; all tests are end to end.
+Tests are run with `./test/test.sh` when no changes are to be expected, and with `./test/test.sh --update-diff` otherwise.
+The diff must be added to the commit that caused it.
+
+Our test suite mostly comprises of characterization tests (a.k.a. snapshot tests, golden master tests, etc.).
```

Comment:
Uber nit, sorry:

```suggestion
Our test suite mostly comprises characterization tests (a.k.a. snapshot tests, golden master tests, etc.).
```

Or "is mostly composed of" might read more easily.

Reply:
"mostly contains" is even more to the point